### ISSUE TITLE
Unneeded printStackTrace

### ISSUE THREAD

**Issue by fbacchella** [zeromq/jeromq#536]
At https://github.com/zeromq/jeromq/blob/713a3e0f0e940497cc58b0d6dcc0f6873e35301e/src/main/java/zmq/Signaler.java#L137

We have  ZMQ that print a stack trace to stderr. I don't think a library should be allowed to directly do that. Especialy when the exception is as quite normal `java.nio.channels.ClosedByInterruptException` and is thrown during . a plain: zpoller.poll(-1L);

**Comment by daveyarwood** [comment]
I agree. If anything, we should be logging it at the debug level, but [logging has been a contentious subject in the past](https://github.com/zeromq/jeromq/issues/128) -- in particular, some people had performance concerns.

**Comment by fbacchella** [comment]
Too much logger.debug(complexObject) ? Then Java 8 plus log4j2 is probably a good solution, see [Log4j 2 and Lambda Expressions](http://www.baeldung.com/log4j-2-lazy-logging) and [Log4j 2's performance](https://logging.apache.org/log4j/2.x/performance.html.)

**Comment by fredoboulo** [comment]
I would be more radical:

- If the exception happens and is handled, there is no need for a stacktrace.
- It the exception happens and is unhandled, that means some devs need to be performed to handle the situation.
- If the exception does not happen, an assert (false) is even better as it enforces the behaviour of the dead branch of the code.

I did not see occurrences of this stacktrace (and the others of this file, but was not keen to enforce it drastically so I left these traces (several of them are in this file as well).

I agree with you: No stacktrace. If the library handles the situation.

**Comment by fbacchella** [comment]
I don't get your explanation. Assertions are not enable by default. So it's for testing only. And what do you mean by "If the exception does not happen" ? If there is no exception, nothing happens, so what the point ?
"I did not see occurrences of this stacktrace". I saw it with code I run, when I interrupt a thread waiting on poll. I didn't found it by luck.
"I agree with you: No stacktrace. If the library handles the situation." No the library must handle the situation. A stack trace to stderr is just lost, it's not in the logfile when an administrator will check for problems, it's not timestamped. It's just noise. So no conditions, just don't.

**Comment by fredoboulo** [comment]
TLDR: These are remnants of devs to check if it was dead code or not. [As per C4](https://rfc.zeromq.org/spec:42/C4), you are welcome to propose a PR.

It seems there is a confusion. These lines are remnant of refactoring process during dev time of bumping to 4.1.7. During that time, the branches of the code where there are these stacktraces were inactive. Thus, stacktraces were introduced to assess if the code was executed, and to avoid breaking production logic. If the conclusion would have been they were active, stacktrace would have been removed. If the conclusion would have been it was actually dead code, assert(false) would have been introduced.

The situation was not definitely settled when the lib was released, so they remained. Thanks for noticing that you have an usage that shows these lines are active.

Regarding assertions, I invite you to read [the coding style](http://zeromq.org/docs:style). There is more in the lore of course!

When I talk about seeing occurrences of this stacktrace: 

- there is no documented usage in the automatic tests that shows that.
- there is no occurrence in the code that I use in production, mostly because my usage of the lib seems to be different from yours.
- there was no report of that occurrence until yours, for a year since this part was rewritten (does not mean it did not occur, of course, just that no-one bothered to raise a flag about it).

Please also have a look at this [discussion about interrupting threads](#116). If you found that using interruptions is a valid case, that's a very nice news!

Anyhow, you can always propose a PR to fix that, the project welcomes all contributions, provided you adhere to the Collective Code Construction Contract (C4).

**Comment by fbacchella** [comment]
A simple use case to generate 3 stack dumps:
```
import java.io.IOException;
import java.nio.channels.SelectableChannel;
import java.nio.channels.Selector;
import java.util.UUID;

import org.zeromq.ZContext;
import org.zeromq.ZMQ;
import org.zeromq.ZMQ.Socket;
import org.zeromq.ZPoller;

public class ZmqBadInterrupt2 {

    public static void main(String[] args) throws IOException, InterruptedException {
        try(ZContext ctx = new ZContext()) {
            Thread t = new Thread() { {
                start();
            }
            @Override
            public void run() {
                try(Socket socket = ctx.createSocket(ZMQ.SUB); Selector selector =  Selector.open(); ZPoller zpoller = new ZPoller(selector)) {
                    socket.setLinger(0);
                    socket.setHWM(1);
                    String iface = "inproc://" + UUID.randomUUID();
                    socket.bind(iface);
                    ZPoller.EventsHandler handler = new ZPoller.EventsHandler() {
                        @Override
                        public boolean events(Socket socket, int events) {
                            System.out.println(events);
                            return true;
                        }

                        @Override
                        public boolean events(SelectableChannel channel, int events) {
                            System.out.println(events);
                            return true;
                        }

                    };
                    zpoller.setGlobalHandler(handler);
                    zpoller.register(socket, handler);
                    System.out.println(zpoller.poll(-1L));
                } catch (Throwable e) {
                    System.out.println(e.getClass());
                }
            }
            };
            Thread.sleep(200);
            t.interrupt();
            Thread.sleep(200);
        };
        System.out.println("closed");
    }

}
```

Even if it's wrong, it's easy to get them, and I should be warned about that in the normal flow of my code.

I got two stack trace on stderr (so generated by the library): from a java.nio.channels.ClosedByInterruptException and from a java.nio.channels.ClosedChannelException.

On stdout, I got:
closed
class java.lang.AssertionError

So I also got an AssertionError, thrown by poll.

**Comment by fredoboulo** [comment]
Hi,

Agreed. We should be highlighting and warning about the misusage of interrupt(). The promise of using interruptions to properly release resources was never fully implemented, AFAIK. I used the latest code and could not get assertions, but program fails to stop from time to time for me, with snapshot or 0.4.3.

To get rid of that interrupt() construct, I would propose an extra socket for control, for example inproc PAIR or PULL, add it to the poller, and stop the polling when a particular message is received .

Or use ZThread.fork, which will directly create and return the control socket (aka pipe) for you.

**Comment by daveyarwood** [comment]
I still see 56 instances of `printStackTrace` in the repo, 18 of which are in the library code (as opposed to in tests, where it might be more feasible to do it).

I agree with @fredoboulo about taking a more extreme approach to this. I don't think there's any reason why we should print stacktraces to STDERR and then continue running the program. We should handle all exceptions or let them explode.
